docker/docker_image_src.go

```diff
@@ -30,15 +32,68 @@ type dockerImageSource struct {
 
 // newImageSource creates a new ImageSource for the specified image reference.
 // The caller must call .Close() on the returned ImageSource.
-func newImageSource(sys *types.SystemContext, ref dockerReference) (*dockerImageSource, error) {
-	c, err := newDockerClientFromRef(sys, ref, false, "pull")
+func newImageSource(ctx context.Context, sys *types.SystemContext, ref dockerReference) (*dockerImageSource, error) {
+	registry, err := sysregistriesv2.FindRegistry(sys, ref.ref.Name())
 	if err != nil {
-		return nil, err
+		return nil, errors.Wrapf(err, "error loading registries")
+	}
+
+	if registry == nil {
+		// No registry was found be the provided reference, so we create a
+		// fallback registry by hand to make the client creation below work as
+		// intended.
+		registry = &sysregistriesv2.Registry{
+			Endpoint: sysregistriesv2.Endpoint{
+				URL: ref.ref.String(),
+			},
+		}
+	}
+
+	// Found the registry within the sysregistriesv2 configuration. Now we test
+	// all endpoints for the manifest availability. If an working image source
+	// was found, it will be used for all future pull actions.
+	var (
+		imageSource     *dockerImageSource
+		manifestLoadErr error
+	)
+	for _, endpoint := range append(registry.Mirrors, registry.Endpoint) {
+		newRef, err := rewriteReference(ref, registry.Prefix, endpoint.URL)
+		if err != nil {
+			return nil, errors.Wrapf(err, "error rewriting reference")
+		}
+
+		client, err := newDockerClientFromRef(sys, newRef, false, "pull")
+		client.endpointSkipVerify = endpoint.Insecure
+		if err != nil {
+			return nil, err
+		}
+
+		testImageSource := &dockerImageSource{
+			ref: newRef,
+			c:   client,
+		}
+
+		manifestLoadErr = testImageSource.ensureManifestIsLoaded(ctx)
+		if manifestLoadErr == nil {
+			imageSource = testImageSource
+			break
+		}
 	}
-	return &dockerImageSource{
-		ref: ref,
-		c:   c,
-	}, nil
+
+	return imageSource, manifestLoadErr
+}
+
+// rewriteReference will substitute the provided reference prefix `from` to the
+// given `to` from the `ref` and creates a new reference from it. The function
+// errors if the newly created reference is not parsable.
+func rewriteReference(ref dockerReference, from, to string) (dockerReference, error) {
+	stringRef := ref.ref.String()
+	if from != "" && strings.HasPrefix(stringRef, from) {
+		newNamedRef := strings.Replace(stringRef, from, to, 1)
+		newParsedRef, err := reference.ParseNamed(newNamedRef)
+		return dockerReference{newParsedRef}, err
```

Comment:
Can we wrap the error and add information that this happened trying to rewrite the reference?

Reply:
> Can we wrap the error and add information that this happened trying to rewrite the reference?

I added an `errors.Wrapf` to the function, which does not rely on the `dockerReference` type any more.

> (Using the official `NewReference` constructor here instead of hard-coding the implementation would also add some peace of mind that the invariants of `dockerReference` will be maintained.)

I cannot use `NewReference` there because it will return an `types.ImageReference`, but I need the concrete `dockerReference` type afterwards.